### PR TITLE
`General`: Open routing via chart in new tab

### PR DESCRIPTION
**Pull request by willergeorg**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Currently, the access of other pages via charts is implemented in that way, that the page changes in the same browser tab. Since the click event is handled by the charting framework, we cannot implement the delegation as link without rebuilding the whole chart from scratch. Especially for charts that provide an overview of different aspects, this might be a suboptimal solution. The user would need to navigate back via browser navigation or breadcrumbs. Instead, I changed the navigation of selected charts in that way, that the new page opens in a separate tab. Then, the user can simply switch tabs to return to the overview.

### Description

I implemented a new routing type in `navigation.utils.ts` that handles the new routing behavior. I replaced the angular router with the auxiliary service within the components.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with submissions
- 1 Programming exercise SCA enabled and submissions that got points deducted by SCA

1. Log in to Artemis as instructor
2. Navigate to the course statistics of the corresponding course
3. Select any bar in the average scores chart. The corresponding exercise statistics should open in a new tab
4. Navigate to the course scores. Selecting any bar should open a new tab with the participation scores view.
5. Navigate to the grading configuration page of the programming exercise. Select any bar part of the points chart in the test cases tab. The behavior should be similar
6. Switch to the SCA tab. Select any part of the lower two bars. Again, the new page should open in a new tab.
7. Navigate to the exam statistics. Select any bar in the score distribution. Make sure that a new tab opens
8. Select any of the horizontal bars in the charts below (only those concerning a specific exercise implement the routing). Here, a new tab should open again.
9. Log out and log in as Student again
10. Navigate to the course statistics tab. Check that the same behavior is observable for both the horizontal bar charts and the line chart.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| `course-scores.component.ts` | 73.88  | 94.72 |
| `exam-scores-average-score-graph.component.ts` | 70.21 | 84.9 |
| `exam-scores.component.ts` | 83.72 | 97.61 |
| `sca-category-distribution-chart.component.ts` | 90.32 | 88.4 |
| `test-case-distribution-chart.component.ts` | 93.33 | 88.46 |
| `exercise-assessment-dashboard.component.ts` | 66.95 | 81.32 |
| `course-statistics.component.ts` | 86.99 | 96.06 |
| `exercise-scores-chart.component.ts` | 79.41 | 98.05 |
| `statistics-average-score-graph.component.ts` | 74.41 | 93.38 |

### Screenshots

Small demo about the intended result:
https://user-images.githubusercontent.com/80622272/160232300-8cb324b6-73dd-4701-ae9d-f903997ed954.mp4

### CODE REVIEW

**Review by filip-gregurevic — DISMISSED** [review]
Changes look good to me, nice work 👍🏼

**Review by Dominik-Weinzierl — APPROVED** [review]
Works as described - Code looks good :+1:

**Review by canberkanar — APPROVED** [review]
Tested on ts2. Tabs open on clicking the bars as described.

File: src/main/webapp/app/course/course-scores/course-scores.component.ts [hunk 1]
```diff
@@ -23,6 +23,8 @@ import { CourseScoresCsvRow, CourseScoresCsvRowBuilder } from 'app/course/course
 import { CourseScoresStudentStatistics } from 'app/course/course-scores/course-scores-student-statistics';
 import { mean, median, standardDeviation } from 'simple-statistics';
 import { ExerciseTypeStatisticsMap } from 'app/course/course-scores/exercise-type-statistics-map';
+import { ChartRoutingService } from 'app/shared/chart/chart-routing.service';
+import { CsvExportOptions } from 'app/shared/export/csv-export-modal.component';
```
Comment (filip-gregurevic) [inline src/main/webapp/app/course/course-scores/course-scores.component.ts:27]:
```suggestion

```
Should probably fix this

**Review by filip-gregurevic — CHANGES_REQUESTED** [review]
This is probably due to resolving some conflicts, I guess this should also fix the build and tests

**Review by Dominik-Weinzierl — APPROVED** [review]
Re-approve :+1: Changes looks good since last approve and tested again on Ts1

**Review by Dominik-Weinzierl — APPROVED** [review]
Re-approve - changes looks good :+1:

File: src/test/javascript/spec/component/assessment-dashboard/exercise-assessment-dashboard.component.spec.ts [hunk 2]
```diff
@@ -319,6 +320,9 @@ describe('ExerciseAssessmentDashboardComponent', () => {
                 comp.submissionsWithComplaints = [submissionWithComplaintDTO];
 
                 accountService = TestBed.inject(AccountService);
+                const navigationUitlService = TestBed.inject(ArtemisNavigationUtilService);
+
+                routingStub = jest.spyOn(navigationUitlService, 'routeInNewTab').mockImplementation();
```
Comment (akesfeden) [inline src/test/javascript/spec/component/assessment-dashboard/exercise-assessment-dashboard.component.spec.ts:325]:
```suggestion
                const navigationUtilService = TestBed.inject(ArtemisNavigationUtilService);

                routingStub = jest.spyOn(navigationUtilService, 'routeInNewTab');
```

As far as I know, `mockImplementation` shouldn't be needed when `MockProvider` is used.

File: src/test/javascript/spec/component/statistics/statistics-average-score-graph.component.spec.ts [hunk 3]
```diff
@@ -39,13 +38,14 @@ describe('StatisticsAverageScoreGraphComponent', () => {
...
-            providers: [{ provide: Router, useClass: MockRouter }],
+            providers: [MockProvider(ArtemisNavigationUtilService)],
         })
             .compileComponents()
             .then(() => {
                 fixture = TestBed.createComponent(StatisticsAverageScoreGraphComponent);
                 component = fixture.componentInstance;
-                router = fixture.debugElement.injector.get(Router);
+                const routingService = TestBed.inject(ArtemisNavigationUtilService);
+                routingStub = jest.spyOn(routingService, 'routeInNewTab').mockImplementation();
```
Comment (akesfeden) [inline src/test/javascript/spec/component/statistics/statistics-average-score-graph.component.spec.ts:48]:
```suggestion
                routingStub = jest.spyOn(routingService, 'routeInNewTab');
```

Same here.

**Review by akesfeden — COMMENTED** [review]
Code changes look good, I just have some minor suggestions.

**Review by cremertim — APPROVED** [review]
Tested on ts3, everything works as expected. Followed the testing steps

**Review by cremertim — APPROVED** [review]
reapporve after local testing

**Review by bassner — APPROVED** [review]
Thanks for adding the changes. LGTM 👍

**Review by Dominik-Weinzierl — APPROVED** [review]
Re-approve after local testing :+1:

**Review by akesfeden — APPROVED** [review]
Thank you for adding the suggested changes, looks good to me!

**Review by filip-gregurevic — APPROVED** [review]
Re-approve 👍🏼

**Pull request merged**